### PR TITLE
Support Mail Server data synchronization

### PR DESCRIPTION
**Pull request by adambabik**
These changes add a support for syncing data between two Mail Servers. It still lacks features that allow triggering syncing easily.

This is more like a preparation to support data syncing between Mail Servers rather than a final implementation. It does not give an easy way to start syncing. This will be solved in the next PR.

Important changes:
- [x] Processing requests by mail servers was refactored.

Todo:
- [x] merged https://github.com/status-im/whisper/pull/13 first.

Relates to #1290

### CODE REVIEW

File: mailserver/mailserver.go [hunk 1]
```diff
@@ -288,24 +290,56 @@ func (s *WMailServer) DeliverMail(peer *whisper.Peer, request *whisper.Envelope)
...
+			if err := s.sendEnvelopes(peer, bundle, batch); err != nil {
+				errCh <- err
+				break
+			}
+		}
+		close(errCh)
+	}()
+
+	start := time.Now()
+	nextPageCursor, lastEnvelopeHash := s.processRequestInBundles(
```
Comment (adambabik) [inline mailserver/mailserver.go:322]: Some stress tests for this fragment would be nice. We use a buffered channel but if there are more than 5 bundles, sending them through the network will become a bottleneck. It is still better than the previous implementation which was sending bundles without batching.

File: mailserver/mailserver_test.go [hunk 2]
```diff
@@ -330,31 +330,41 @@ func (s *MailserverSuite) TestRequestPaginationLimit() {
 	s.Nil(cursor)
 	s.Equal(params.limit, limit)
 
-	envelopes, _, cursor, err := s.server.processRequest(nil, lower, upper, bloom, limit, nil, false)
-	s.NoError(err)
-	for _, env := range envelopes {
-		receivedHashes = append(receivedHashes, env.Hash())
+	iter := s.server.createIterator(lower, upper, cursor)
```
Comment (adambabik) [inline mailserver/mailserver_test.go]: This fragment repeats in a few places. It can be extracted to a method.

**Review by gravityblast — APPROVED** [review]

File: Gopkg.lock [hunk 3]
```diff
@@ -822,12 +822,12 @@
   revision = "fbcc46a78cd43fef95a110df664aab513116a850"
 
 [[projects]]
-  digest = "1:6cb252f27feb57ef0e8406556c259d903c0ecff2ab0d2200ca85773b3561777d"
+  branch = "sync-messages"
```
Comment (gravityblast) [inline Gopkg.lock]: just as reminder, we have to remember to change this after the PR on whisper is merged

File: mailserver/mailserver.go [hunk 4]
```diff
@@ -288,24 +290,56 @@ func (s *WMailServer) DeliverMail(peer *whisper.Peer, request *whisper.Envelope)
...
 		requestsBatchedCounter.Inc(1)
 	}
 
-	_, lastEnvelopeHash, nextPageCursor, err := s.processRequest(
-		peer,
-		lower, upper,
+	iter := s.createIterator(lower, upper, cursor)
+	defer iter.Release()
+
+	bundles := make(chan []*whisper.Envelope, 5)
```
Comment (dshulyak) [inline mailserver/mailserver.go:308]: is it just to avoid blocking or this number means something else?

**Review by dshulyak — APPROVED** [review]

File: mailserver/mailserver.go (hunk 4 above)
Comment (adambabik) [reply mailserver/mailserver.go:308]: Just avoid blocking. Pretty arbitrary.

**Pull request merged**